[Refactor] Make JSDebugger component handle display logic previously in DebugArea class

pcardune opened the pull request:
All the js debugger code is kind of a mix of React and non-react. We were using react to do the initial dom render (as if it were an ejs template) and then using jquery to modify that dom. This PR starts to change that by doing the following:

* Move related jsdebugger files into a single directory under lib.
* Remove `DebugArea` completely and put it's logic into `JsDebugger` directly.
* Replace jQuery.animate calls with css transitions
* Add unit tests for the logic previously covered by DebugArea.

Bjvanminnen commented:
Would be nice to have some description provided.

pcardune commented:
@Bjvanminnen updated. For some reason I have a bad habit of assigning PRs to people before updating the description :(. But it's updated now!

Bjvanminnen commented on apps/src/applab/applab.js:
```diff
@@ -782,7 +776,8 @@ Applab.init = function (config) {
     showDebugSlider: showDebugConsole,
-    showDebugWatch: config.level.showDebugWatch || experiments.isEnabled('showWatchers')
+    showDebugWatch: config.level.showDebugWatch || experiments.isEnabled('showWatchers'),
+    debuggerUi,
```
All of our other pageConstants are (I believe) simple bools or strings. This sort of seems like it should be a separate concept.

It actually feels a little weird to me to have this be in our redux store in general. It isn't really state, right? It's a bunch of code.

pcardune replied:
Sadly, it does have state. It holds a command history and an instance of the js interpreter.

At the moment it does not use redux for any of these, so the easiest way to get at the JsDebuggerUi instance without having to thread it through all the components, and without using other globals like the Applab or Gamelab singletons, is to put it into redux. pageConstants is not necessarily the right place for it though, and IMO, it should eventually be refactored to just store it's state in redux with it's own set of reducers and whatnot.

I could do that refactoring now as part of this PR. Or I could do it in a separate PR. Do you have a preference?

Bjvanminnen replied:
I'm fine with that being in a separate PR, but then might recommend a short comment explaining that this doesn't really fit the mold for pageConstants, and we intend to move it in the nearish future.

islemaster replied:
I'm still confused about why this needs to be in Redux - so that we can hide it (switch to design mode) and re-render it with state when we switch back?  The command history definitely seems like it should be stored in redux by itself.  The JsInterpreter instance, I think, is only created when we execute a command by cloning the main JSInterpreter instance and running the one statement - I'm not sure we need to keep it around.

pcardune replied:
@islemaster then it's really just for the command history.

Bjvanminnen commented on apps/src/lib/jsdebugger/JsDebugger.jsx:
```diff
@@ -143,10 +152,60 @@ var UnconnectedJsDebugger = React.createClass({
+  },
+
+  slideShut() {
+    const closedHeight = $(this.root).find('#debug-area-header').height() + $(this.root).find('#debugResizeBar').height();
```
nit: line is too long

Bjvanminnen commented on apps/src/lib/jsdebugger/JsDebugger.jsx:
```diff
@@ -143,10 +152,60 @@ var UnconnectedJsDebugger = React.createClass({
+      openedHeight: $(this.root).height(),
+      closedHeight,
+    });
+    $('#codeTextbox').animate(
```
For cases like this, where we're modifying portions of the DOM that we don't own, it might be worth a short comment explaining where this comes from.

I wonder if we could/should grab a reference to `#codeTextbox` on mount, comment there, and then elsewhere do something like `this.codeTextbox`.

islemaster replied:
Alternatively I wonder if this animation logic touching both the debugger and the editor areas should live outside of either.  "Collapsing pane" feels like a separate concern that should be handled outside of this component.

Bjvanminnen replied:
PR description says you're replacing jquery animate calls with CSS transitions? (isn't this a jquery animate call?)

pcardune replied:
Yeah I thought about that. But at the same time, if you don't know about the lifecycle of the dom node, you could end up referring to a dom node that no longer exists. Perhaps it would be better to invert the control here and pass in a `onResize` prop to `JsDebugger` which did this animation bit?

pcardune replied:
I replaced other animate calls. This one still needs to get moved to a css transition, but that should really happen in a different component.

Bjvanminnen commented on apps/src/lib/jsdebugger/JsDebugger.jsx:
```diff
@@ -170,7 +243,11 @@ var UnconnectedJsDebugger = React.createClass({
+            className={`fa ${this.state.open ? 'fa-chevron-circle-down' : 'fa-chevron-circle-up'}`}
+            style={styles.showHideIcon}
+            onClick={this.slideToggle}
+          />
```
Can we use FontAwesome component?

Bjvanminnen commented on apps/test/unit/lib/jsdebugger/JsDebuggerTest.js:
```diff
@@ -0,0 +1,122 @@
+
+    const codeTextbox = document.createElement('div');
+    document.body.appendChild(codeTextbox);
+    codeTextbox.id = 'codeTextbox';
```
Should we remove this from the DOM after each? Right now you're adding one per test - is the component always going to use the most recently added one, or will it use an element from a previous test?

Bjvanminnen commented on apps/test/unit/lib/jsdebugger/JsDebuggerTest.js:
```diff
@@ -0,0 +1,122 @@
+  const debugAreaEl = () => root.find('#debug-area').get(0);
+  const paneHeader = () => root.find('PaneHeader');
+  const closeIcon = () => paneHeader().find('i.fa-chevron-circle-down');
+  const openIcon = () => paneHeader().find('i.fa-chevron-circle-up');
```
why not have these be in the beforeEach? (similar to how you do jsDebugger)

pcardune replied:
the const definitions are not in beforeEach because they never change. And they are functions because their return values might change depending on what actions have been taken. For example:

```jsx
closeIcon(); // returns a ReactWrapper with the close icon inside it.
closeIcon().simulate('click'); // changes state which makes the closeIcon go away
closeIcon(); // returns an empty ReactWrapper since the closeIcon is no longer there
```

if I instead has a `closeIcon` variable at the top level which wasn't a function, then i'd have to remember to reset it's value after calling `closeIcon.simulate('click')`, which would be more error prone.

Bjvanminnen commented on apps/test/unit/lib/jsdebugger/JsDebuggerTest.js:
```diff
@@ -0,0 +1,122 @@
+    });
+
+  });
+
```
why all these extra lines of whitespace?

pcardune replied:
easier to read?

islemaster commented on apps/src/lib/jsdebugger/JsDebuggerUi.js:
```diff
@@ -1,13 +1,12 @@
+var Observer = require('@cdo/apps/Observer');
+var utils = require('@cdo/apps/utils');
+import {setStepSpeed} from '@cdo/apps/redux/runState';
+import {add, update, remove} from '@cdo/apps/redux/watchedExpressions';
```
I've been avoiding the `@cdo/apps` alias in regular source code because my refactoring tools don't handle it automatically like they do for relative paths.  Do you know of better tools for this?  Also, any reason not to change all of these to ES6 imports?

pcardune replied:
I do not know of better tools 🙁 . I'll try to stick to relative imports inside `src`. I didn't know anyone was using any refactoring tools.

islemaster replied:
JetBrains (esp. Webstorm) move/rename refactoring tools for JS are pretty good!

islemaster commented:
Shouldn't `apps/src/lib/jsdebugger` be `apps/src/lib/tools/jsdebugger`?

islemaster commented:
Huge improvement in test coverage of this feature :clap: thanks for this!

pcardune commented:
Ok, addressed all the comments except for where/why we are storing JsDebuggerUi inside redux. I'd like to have that change get addressed in a separate PR if it's alright with y'alls.

Bjvanminnen commented on apps/src/lib/tools/jsdebugger/JsDebugger.jsx:
```diff
@@ -3,19 +3,22 @@
+import styleConstants from '../../../styleConstants';
+import {ConnectedWatchers} from '../../../templates/watchers/Watchers';
+import PaneHeader from '../../../templates/PaneHeader';
+const {PaneSection, PaneButton} = PaneHeader;
```
This could also be `import PaneHeader, { PaneSection, PaneButton } from '../../../templates/PaneHeader'`

pcardune replied:
sadly, not. It looks like this in Paneheader.jsx: 😢 😢 

```jsx
module.exports = PaneHeader;
PaneHeader.PaneSection = PaneSection;
PaneHeader.PaneButton = PaneButton;
```

Bjvanminnen commented on apps/src/lib/tools/jsdebugger/JsDebugger.jsx:
```diff
@@ -125,28 +129,82 @@ var DebugButtons = function () {
+
+  slideShut() {
+    const closedHeight = $(this.root).find('#debug-area-header').height() +
+                         $(this.root).find('#debugResizeBar').height();
```
nit: should only be indented two spaces beyond line 177
